pyro/distributions/transforms/affine_autoregressive.py

```diff
@@ -220,6 +221,111 @@ def _inverse_stable(self, y):
+    >>> z = torch.rand(batch_size, context_dim)
+    >>> flow_dist = dist.ConditionalTransformedDistribution(base_dist,
+    ... [transform]).condition(z)
+    >>> flow_dist.sample(sample_shape=torch.Size([batch_size]))  # doctest: +SKIP
+
+    The inverse of the Bijector is required when, e.g., scoring the log density of a
+    sample with :class:`~pyro.distributions.TransformedDistribution`. This
+    implementation caches the inverse of the Bijector when its forward operation is
+    called, e.g., when sampling from
+    :class:`~pyro.distributions.TransformedDistribution`. However, if the cached
+    value isn't available, either because it was overwritten during sampling a new
+    value or an arbitary value is being scored, it will calculate it manually. Note
+    that this is an operation that scales as O(D) where D is the input dimension,
+    and so should be avoided for large dimensional uses. So in general, it is cheap
+    to sample from IAF and score a value that was sampled by IAF, but expensive to
+    score an arbitrary value.
+
+    :param autoregressive_nn: an autoregressive neural network whose forward call
+        returns a real-valued mean and logit-scale as a tuple
+    :type autoregressive_nn: nn.Module
+    :param log_scale_min_clip: The minimum value for clipping the log(scale) from
+        the autoregressive NN
+    :type log_scale_min_clip: float
+    :param log_scale_max_clip: The maximum value for clipping the log(scale) from
+        the autoregressive NN
+    :type log_scale_max_clip: float
+    :param sigmoid_bias: A term to add the logit of the input when using the stable
+        tranform.
+    :type sigmoid_bias: float
+    :param stable: When true, uses the alternative "stable" version of the transform
+        (see above).
+    :type stable: bool
+
+    References:
+
+    [1] Diederik P. Kingma, Tim Salimans, Rafal Jozefowicz, Xi Chen, Ilya Sutskever,
+    Max Welling. Improving Variational Inference with Inverse Autoregressive Flow.
+    [arXiv:1606.04934]
+
+    [2] Danilo Jimenez Rezende, Shakir Mohamed. Variational Inference with
+    Normalizing Flows. [arXiv:1505.05770]
+
+    [3] Mathieu Germain, Karol Gregor, Iain Murray, Hugo Larochelle. MADE: Masked
+    Autoencoder for Distribution Estimation. [arXiv:1502.03509]
+
+    """
+
+    domain = constraints.real
+    codomain = constraints.real
+    bijective = True
+    event_dim = 1
+
+    def __init__(self, autoregressive_nn, **kwargs):
+        super().__init__()
+        self.nn = autoregressive_nn
+        self.kwargs = kwargs
+
+    def condition(self, context):
+        """
+        Conditions on a context variable, returning a non-conditional transform of
+        of type :class:`~pyro.distributions.transforms.AffineAutoregressive`.
+        """
+
+        # Note that nn.condition doesn't copy the weights of the ConditionalAutoregressiveNN
+        return AffineAutoregressive(self.nn.condition(context), **self.kwargs)
```

Comment:
@fritzo Can you please review this line? I have made a design change here. Previously I was using `partial` to fix the context that is used in the `ConditionalAutoRegressiveNN`, but this resulted in some bugs... You need to pass an `nn.Module` class to `AffineAutoregressive` rather than a function completion so that it registers the parameters and you can inspect the properties inside like the permutation (used in the tests)

Reply:
I think in a different programming language, I would subclass `ConditionalAutoRegressiveNN` to include a context property, overwrite `.forward` to use this context, then copy self.nn into a new object of this type. That way you could share the weights between multiple `self.nn.condition(context)` objects rather than have the context rewritten each time it is called

Reply:
@fritzo @stefanwebb have you guys resolved this design issue? otherwise lgtm

Reply:
Sorry I let this drop. I don't fully understand the design issue. Specifically what contract is made by `nn.condition()`? Does it always return a neural net? Which of `nn` and `nn.condition(...)` is ever serialized or trained? What bugs were caused by unregistered parameters? @stefanwebb let me know if you'd like to zoom to discuss any time.